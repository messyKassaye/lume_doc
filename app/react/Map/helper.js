const _mergeArrays = (first, second) => {
  let bigArray;
  let smallArray;
  if (first.length > second.length) {
    bigArray = first;
    smallArray = second;
  } else {
    bigArray = second;
    smallArray = first;
  }

  smallArray.forEach(item => bigArray.push(item));
  return bigArray;
};

const getMarkersBoudingBox = markers =>
  markers.reduce(
    (_b, marker) => {
      const b = _b.slice();
      if (!b[0][0] || marker.longitude < b[0][0]) {
        b[0][0] = marker.longitude;
      }
      if (!b[1][0] || marker.longitude > b[1][0]) {
        b[1][0] = marker.longitude;
      }
      if (!b[1][1] || marker.latitude > b[1][1]) {
        b[1][1] = marker.latitude;
      }
      if (!b[0][1] || marker.latitude < b[0][1]) {
        b[0][1] = marker.latitude;
      }
      return b;
    },
    [
      [null, null],
      [null, null],
    ]
  );

const markersToStyleFormat = markers =>
  markers.map((marker, index) => {
    const properties = marker.properties || {};
    const { longitude, latitude } = marker;
    properties.index = index;
    return {
      type: 'Feature',
      properties,
      geometry: {
        type: 'Point',
        coordinates: [longitude, latitude],
      },
    };
  });

const noop = () => {};

const extractPropertyBasedMarkers = (entity, geolocationProp, propertyMetadata, color) => {
  const markers = [];
  propertyMetadata
    .filter(mo => mo.get('value'))
    .forEach(point => {
      markers.push({
        properties: { entity: entity.toJS(), color, info: point.getIn(['value', 'label']) },
        latitude: point.getIn(['value', 'lat']),
        longitude: point.getIn(['value', 'lon']),
        label: geolocationProp.get('label'),
      });
    });
  return markers;
};

const extractInheritedMarkers = (entity, geolocationProp, propertyMetadata, color) => {
  const markers = [];
  propertyMetadata.forEach(relatedProperty => {
    if (relatedProperty.has('inherit_geolocation')) {
      // conditional is a quick hack to prevent app crash
      relatedProperty
        .get('inherit_geolocation')
        .filter(mo => mo.get('value'))
        .forEach(point => {
          markers.push({
            properties: {
              entity: entity.toJS(),
              color,
              info: point.getIn(['value', 'label']),
              inherited: true,
              label: relatedProperty.get('label'),
              context: geolocationProp.get('content'),
            },
            latitude: point.getIn(['value', 'lat']),
            longitude: point.getIn(['value', 'lon']),
            label: geolocationProp.get('label'),
          });
        });
    }
  });
  return markers;
};

const extractMarkers = ({ entity, geolocationProps, metadata, color, type }) => {
  let markers = [];
  const geolocationPropNames = geolocationProps.map(prop => prop.get('name'));

  if (geolocationProps.size && metadata) {
    const [...properties] = metadata.keys();
    properties.forEach(property => {
      if (geolocationPropNames.includes(property) && metadata.has(property)) {
        const geolocationProp = geolocationProps.find(p => p.get('name') === property);
        const propertyMetadata = metadata.get(property);
        if (type === 'PropertyBased') {
          markers = _mergeArrays(
            markers,
            extractPropertyBasedMarkers(entity, geolocationProp, propertyMetadata, color)
          );
        } else {
          markers = _mergeArrays(
            markers,
            extractInheritedMarkers(entity, geolocationProp, propertyMetadata, color)
          );
        }
      }
    });
  }
  return markers;
};

const getPropertyBasedMarkers = (template, entity) => {
  const color = template.get('color');
  const templateProperties = template.get('properties');
  const metadata = entity.get('metadata');

  const geolocationProps = templateProperties.filter(
    property => property.get('type') === 'geolocation'
  );

  return extractMarkers({ entity, geolocationProps, metadata, color, type: 'PropertyBased' });
};

const getInheritedMarkers = (template, entity, templates) => {
  const templateProperties = template.get('properties');
  const color = template.get('color');
  const metadata = entity.get('metadata');

  const geolocationProps = templateProperties.filter(property => {
    if (property.get('type') === 'relationship' && property.has('inheritProperty')) {
      const contentTemplate = templates.find(
        t => t.get('_id').toString() === property.get('content').toString()
      );
      const inheritedProperty = contentTemplate
        .get('properties')
        .find(p => p.get('_id').toString() === property.get('inheritProperty').toString());
      return inheritedProperty.get('type') === 'geolocation';
    }
    return false;
  });

  return extractMarkers({ entity, geolocationProps, metadata, color, type: 'Inherited' });
};

const getEntityMarkers = (entity, templates) => {
  const template = templates.find(_t => _t.get('_id') === entity.get('template'));

  const propertyBasedMarkers = getPropertyBasedMarkers(template, entity);
  const inheritedMarkers = getInheritedMarkers(template, entity, templates);

  return _mergeArrays(propertyBasedMarkers, inheritedMarkers);
};

const getMarkers = (entities, templates) => {
  const m = entities.reduce((markers, entity) => {
    const entityMarkers = getEntityMarkers(entity, templates);
    return _mergeArrays(markers, entityMarkers);
  }, []);
  return m;
};

const TRANSITION_PROPS = {
  transitionDuration: 0,
  transitionEasing: t => t,
  transitionInterruption: 1,
  onTransitionStart: noop,
  onTransitionInterrupt: noop,
  onTransitionEnd: noop,
  onViewportChange: noop,
  onStateChange: noop,
};

export { getMarkersBoudingBox, markersToStyleFormat, getMarkers, TRANSITION_PROPS };
