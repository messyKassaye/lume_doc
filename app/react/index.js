import React from 'react';
import ReactDOM from 'react-dom';

import { AppContainer } from 'react-hot-loader';

import App from './App';

import './App/sockets';

const render = Component => {
  ReactDOM.hydrate(
    <AppContainer>
      <Component />
    </AppContainer>,
    document.getElementById('root')
  );
};

render(App);

if (module.hot) {
  module.hot.accept('./App', () => {
    const nextApp = require('./App');
    render(nextApp);
  });
}
