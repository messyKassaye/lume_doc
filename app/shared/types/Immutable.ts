import Immutable from 'immutable';

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? Immutable.List<IImmutable<Elem>>
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
