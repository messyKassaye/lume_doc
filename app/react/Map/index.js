import * as helper from './helper';

export { default as Map } from './Map.js';
export { default as Markers } from './Markers.js';
export { helper };
