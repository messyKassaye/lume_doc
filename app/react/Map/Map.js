import PropTypes from 'prop-types';
import React, { Component } from 'react';
import ReactMapGL, { Marker, Popup, NavigationControl, setRTLTextPlugin } from 'react-map-gl';
import Immutable from 'immutable';
import { Icon } from 'UI';
import Supercluster from 'supercluster'; //eslint-disable-line
import settingsAPI from 'app/Settings/SettingsAPI';
import _style from './style.json';
import { getMarkersBoudingBox, markersToStyleFormat, TRANSITION_PROPS } from './helper';

setRTLTextPlugin(
  'https://api.mapbox.com/mapbox-gl-js/plugins/mapbox-gl-rtl-text/v0.2.3/mapbox-gl-rtl-text.js',
  null, //callback
  true //lazy load
);

const PRIME_MERIDIAN = 0;

const getStateDefaults = ({ latitude, longitude, width, height, zoom }) => ({
  viewport: {
    latitude: latitude || PRIME_MERIDIAN,
    longitude: longitude || PRIME_MERIDIAN,
    width: width || 250,
    height: height || 200,
    zoom,
  },
  selectedMarker: null,
  settings: { scrollZoom: true, touchZoom: true },
  showControls: false,
});

export default class Map extends Component {
  constructor(props) {
    super(props);
    this.state = getStateDefaults(props);
    this.state.settings.scrollZoom = props.scrollZoom;
    this.state.settings.touchZoom = props.scrollZoom;
    this.state.showControls = props.showControls;

    this.mapStyle = Immutable.fromJS(_style);
    this.supercluster = new Supercluster({
      radius: _style.sources.markers.clusterRadius,
      maxZoom: _style.sources.markers.clusterMaxZoom,
    });

    this.updateMapStyle(props);
    this.bindActions();
    this.assignDefaults();
  }

  async componentDidMount() {
    this.collectionSettings = await settingsAPI.get();
    this.setDefaultCoordinates();

    this.replaceKeysMapStyleJson();

    this.setSize();
    const map = this.map && this.map.getMap();
    if (map) {
      map.on('load', () => this.centerOnMarkers(this.props.markers));
      map.on('moveend', e => {
        if (e.autoCentered) {
          this.setViewport(map);
        }
      });
    }
    this.eventListener = window.addEventListener('resize', this.setSize);
  }

  componentWillReceiveProps(props) {
    const { viewport } = this.state;
    const { markers } = props;
    const latitude = props.latitude || viewport.latitude;
    const longitude = props.longitude || viewport.longitude;
    const newViewport = { ...viewport, latitude, longitude, markers };

    if (!Immutable.fromJS(this.props.markers).equals(Immutable.fromJS(markers))) {
      this.autoCenter(markers);
      this.updateMapStyle(props);
    }
    this.setState({ viewport: newViewport });
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.setSize);
  }

  onClick(e) {
    const { markers, onClick } = this.props;
    const unclusteredPoints = e.features.filter(f => f.layer.id === 'unclustered-point');
    const cluster = e.features.find(f => f.layer.id === 'clusters');

    switch (true) {
      case unclusteredPoints.length === 1:
        this.clickOnMarker(markers[unclusteredPoints[0].properties.index]);
        break;
      case unclusteredPoints.length > 1:
        this.clickOnCluster(unclusteredPoints.map(marker => markers[marker.properties.index]));
        break;
      default:
        break;
    }

    if (cluster) {
      this.processClusterOnClick(cluster);
    }

    onClick(e);
  }

  onHover(e) {
    const { markers, cluster } = this.props;
    const { selectedMarker } = this.state;
    let feature = null;

    if (e.features) {
      feature = e.features.find(f => f.layer.id === 'unclustered-point');
      if (feature) {
        this.hoverOnMarker(markers[feature.properties.index]);
      }
    }
    if (!feature && selectedMarker && cluster) {
      this.setState({ selectedMarker: null });
    }
  }

  setDefaultCoordinates() {
    const { viewport } = this.state;
    if (viewport.latitude === PRIME_MERIDIAN && viewport.longitude === PRIME_MERIDIAN) {
      viewport.latitude = this.collectionSettings.mapStartingPoint[0].lat;
      viewport.longitude = this.collectionSettings.mapStartingPoint[0].lon;
      this.setState({ viewport });
    }
  }

  setSize() {
    const { viewport } = this.state;
    const { width, height } = this.props;
    viewport.width = width || (this.container && this.container.offsetWidth);
    viewport.height = height || (this.container && this.container.offsetHeight);
    this.setState({ viewport });
  }

  setViewport(map) {
    const { viewport } = this.state;
    const newViewport = Object.assign(viewport, {
      latitude: map.getCenter().lat,
      longitude: map.getCenter().lng,
      zoom: map.getZoom(),
      pitch: map.getPitch(),
      bearing: map.getBearing(),
    });
    this.setState({ viewport: newViewport });
  }

  bindActions() {
    this.zoom = this.zoom.bind(this);
    this.setSize = this.setSize.bind(this);
    this.onClick = this.onClick.bind(this);
    this.onHover = this.onHover.bind(this);
  }

  assignDefaults() {
    this._onViewportChange = viewport => {
      this.setState({ viewport });
    };
    this._onViewStateChange = viewport => {
      this.setState({ viewport });
    };

    this.zoomIn = () => this.zoom(+1);
    this.zoomOut = () => this.zoom(-1);

    this.interactiveLayerIds = [];
    if (this.mapStyle.has('layers')) {
      this.interactiveLayerIds = this.mapStyle
        .get('layers')
        .filter(l => l.get('interactive'))
        .map(l => l.get('id'))
        .toJS();
    }
  }

  processClusterOnClick(cluster) {
    const currentData = this.mapStyle.getIn(['sources', 'markers', 'data', 'features']).toJS();
    this.supercluster.load(currentData);
    const markersOnCluster = this.supercluster.getLeaves(cluster.properties.cluster_id, Infinity);
    this.clickOnCluster(markersOnCluster);
  }

  autoCenter(markers) {
    const { autoCenter } = this.props;
    if (autoCenter) {
      this.centerOnMarkers(markers);
    }
  }

  centerOnMarkers(markers) {
    if (!this.map || !markers.length) {
      return;
    }
    const map = this.map.getMap();
    const boundaries = getMarkersBoudingBox(markers);
    map
      .stop()
      .fitBounds(
        boundaries,
        { padding: { top: 70, left: 20, right: 20, bottom: 20 }, maxZoom: 5, duration: 0 },
        { autoCentered: true }
      );
  }

  zoom(amount) {
    const { viewport } = this.state;
    if (!this.map) {
      return;
    }
    const map = this.map.getMap();
    this._onViewStateChange({
      ...viewport,
      zoom: map.getZoom() + amount,
      ...TRANSITION_PROPS,
      transitionDuration: 500,
    });
  }

  updateMapStyle({ cluster, markers }) {
    if (!cluster) {
      return;
    }

    const markersData = markersToStyleFormat(markers);
    const currentData = this.mapStyle.getIn(['sources', 'markers', 'data', 'features']);
    if (!currentData.equals(Immutable.fromJS(markersData))) {
      const style = this.mapStyle.setIn(
        ['sources', 'markers', 'data', 'features'],
        Immutable.fromJS(markersData)
      );
      this.mapStyle = style;
      this.supercluster.load(markersData);
    }
  }

  clickOnMarker(marker) {
    const { clickOnMarker } = this.props;
    clickOnMarker(marker);
  }

  clickOnCluster(cluster) {
    const { clickOnCluster } = this.props;
    clickOnCluster(cluster);
  }

  hoverOnMarker(marker) {
    const { selectedMarker } = this.state;
    const { hoverOnMarker } = this.props;

    if (selectedMarker !== marker) {
      this.setState({ selectedMarker: marker });
    }

    hoverOnMarker(marker);
  }

  replaceKeysMapStyleJson() {
    const { mapTilerKey } = this.collectionSettings;
    const mapTilerKeyPlaceholder = /{{MAP_TILER_KEY}}/g;
    const stringifyStyle = JSON.stringify(this.mapStyle).replace(
      mapTilerKeyPlaceholder,
      mapTilerKey
    );
    this.mapStyle = Immutable.fromJS(JSON.parse(stringifyStyle));
    this.setState({ mapStyleLoaded: true });
  }

  renderMarker(marker, onClick, onMouseEnter, onMouseLeave) {
    const { renderMarker } = this.props;
    if (renderMarker) {
      return renderMarker(marker, onClick);
    }
    return (
      <Icon
        style={{
          position: 'relative',
          top: '-25px',
          right: '15px',
          color: (marker.properties && marker.properties.color) || '#d9534e',
        }}
        icon="map-marker"
        size="2x"
        fixedWidth
        onClick={onClick}
        onMouseOver={onMouseEnter}
        onMouseLeave={onMouseLeave}
      />
    );
  }

  renderPopup() {
    const { selectedMarker } = this.state;
    const { renderPopupInfo } = this.props;
    if (
      selectedMarker &&
      (renderPopupInfo ||
        (selectedMarker && selectedMarker.properties && selectedMarker.properties.info))
    ) {
      const { longitude, latitude } = selectedMarker;
      return (
        <Popup
          tipSize={6}
          anchor="bottom"
          longitude={longitude}
          latitude={latitude}
          offsetTop={-20}
          closeButton={false}
        >
          <div>
            {renderPopupInfo ? renderPopupInfo(selectedMarker) : selectedMarker.properties.info}
          </div>
        </Popup>
      );
    }
    return false;
  }

  renderMarkers() {
    const { markers, cluster } = this.props;
    if (cluster) {
      return false;
    }

    return markers.map((marker, index) => {
      const onClick = this.clickOnMarker.bind(this, marker);
      const onMouseEnter = this.hoverOnMarker.bind(this, marker);
      const onMouseLeave = this.hoverOnMarker.bind(this, null);
      return (
        <Marker {...marker} key={index} offsetLeft={0} offsetTop={0}>
          {this.renderMarker(marker, onClick, onMouseEnter, onMouseLeave)}
        </Marker>
      );
    });
  }

  renderControls() {
    const { showControls } = this.state;
    if (showControls) {
      return (
        <div className="mapbox-navigation">
          <NavigationControl onViewportChange={this._onViewportChange} />
        </div>
      );
    }

    return false;
  }

  render() {
    const { viewport, settings } = this.state;

    return (
      <div
        className="map-container"
        ref={container => {
          this.container = container;
        }}
        style={{ width: '100%', height: '100%' }}
      >
        {this.state && this.state.mapStyleLoaded && (
          <ReactMapGL
            ref={ref => {
              this.map = ref;
            }}
            {...viewport}
            {...settings}
            dragRotate
            mapStyle={this.mapStyle}
            onViewportChange={this._onViewportChange}
            onViewStateChange={this._onViewStateChange}
            onClick={this.onClick}
            onHover={this.onHover}
            interactiveLayerIds={this.interactiveLayerIds}
          >
            {this.renderMarkers()}
            {this.renderPopup()}
            {this.renderControls()}
            <span className="mapbox-help">
              <Icon icon="question-circle" />
              <span className="mapbox-tooltip">Hold shift to rotate the map</span>
            </span>
          </ReactMapGL>
        )}
      </div>
    );
  }
}

Map.defaultProps = {
  markers: [],
  latitude: null,
  longitude: null,
  zoom: 4,
  width: null,
  height: null,
  onClick: () => {},
  clickOnMarker: () => {},
  hoverOnMarker: () => {},
  clickOnCluster: () => {},
  renderPopupInfo: null,
  renderMarker: null,
  cluster: false,
  autoCenter: true,
  scrollZoom: true,
  showControls: false,
  interactiveLayerIds: [],
};

Map.propTypes = {
  markers: PropTypes.arrayOf(PropTypes.object),
  latitude: PropTypes.number,
  longitude: PropTypes.number,
  zoom: PropTypes.number,
  width: PropTypes.number,
  height: PropTypes.number,
  onClick: PropTypes.func,
  clickOnMarker: PropTypes.func,
  renderPopupInfo: PropTypes.func,
  clickOnCluster: PropTypes.func,
  hoverOnMarker: PropTypes.func,
  renderMarker: PropTypes.func,
  cluster: PropTypes.bool,
  autoCenter: PropTypes.bool,
  scrollZoom: PropTypes.bool,
  showControls: PropTypes.bool,
  interactiveLayerIds: PropTypes.arrayOf(PropTypes.string),
};
