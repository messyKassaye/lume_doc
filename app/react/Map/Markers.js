import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Immutable from 'immutable';
import { createSelector } from 'reselect';
import { getMarkers } from './helper';

const selectMarkers = createSelector(
  ({ entities }) => entities,
  ({ state }) => state.templates,
  (entities, templates) => getMarkers(entities, templates)
);

export const MarkersComponent = ({ children, markers }) => children(markers);

MarkersComponent.defaultProps = {
  entities: Immutable.List(),
};

MarkersComponent.propTypes = {
  children: PropTypes.func.isRequired,
  templates: PropTypes.instanceOf(Immutable.List).isRequired,
  entities: PropTypes.instanceOf(Immutable.List),
};

export const mapStateToProps = (state, ownProps) => ({
  markers: selectMarkers({ state, entities: ownProps.entities }),
  templates: state.templates,
});

export default connect(mapStateToProps)(MarkersComponent);
