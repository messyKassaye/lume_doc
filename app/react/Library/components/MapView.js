import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Immutable from 'immutable';
import { Map, Markers } from 'app/Map';
import { bindActionCreators } from 'redux';
import { wrapDispatch } from 'app/Multireducer';
import {
  getAndSelectDocument,
  selectDocuments,
  unselectAllDocuments,
} from 'app/Library/actions/libraryActions';
import SearchBar from 'app/Library/components/SearchBar';
import { TemplateLabel } from 'app/Layout';
import { t } from 'app/I18N';
import { Icon } from 'app/UI';

const markerInfo = info => (
  <div className="marker-info">
    <Icon className="tag-icon" icon="tag" />
    {info}
  </div>
);

export const renderInfo = marker => (
  <div className="popup-container">
    <div className="template-label">
      <TemplateLabel template={marker.properties.entity.template} />
    </div>
    <div className="entity-data">
      <div>
        <span className="popup-name">{marker.properties.entity.title}</span>
        <span className="popup-metadata-property">
          ({t(marker.properties.entity.template, marker.label)})
        </span>
      </div>
      {marker.properties.inherited &&
        markerInfo(<span className="entity-title">{marker.properties.label}</span>)}
      {marker.properties.info && !marker.properties.inherited && markerInfo(marker.properties.info)}
    </div>
  </div>
);

export class MapView extends Component {
  constructor(props) {
    super(props);
    this.clickOnMarker = this.clickOnMarker.bind(this);
    this.clickOnCluster = this.clickOnCluster.bind(this);
  }

  clickOnMarker(marker) {
    this.props.getAndSelectDocument(marker.properties.entity.sharedId);
  }

  clickOnCluster(cluster) {
    this.props.unselectAllDocuments();
    this.props.selectDocuments(cluster.map(m => m.properties.entity));
  }

  render() {
    const { storeKey, markers } = this.props;
    return (
      <div className="library-map main-wrapper" style={{ width: '100%', height: '100%' }}>
        <div className="search-list">
          <SearchBar storeKey={storeKey} />
        </div>
        <div className="documents-counter">
          <span>
            <b>{markers.get('totalRows')}</b> {t('System', 'documents')}
          </span>
        </div>
        <Markers entities={markers.get('rows')}>
          {processedMarkers => (
            <Map
              ref={ref => {
                this.map = ref;
              }}
              markers={processedMarkers}
              zoom={1}
              clickOnMarker={this.clickOnMarker}
              clickOnCluster={this.clickOnCluster}
              renderPopupInfo={renderInfo}
              cluster
            />
          )}
        </Markers>
      </div>
    );
  }
}

MapView.propTypes = {
  markers: PropTypes.instanceOf(Immutable.Map).isRequired,
  storeKey: PropTypes.string.isRequired,
  getAndSelectDocument: PropTypes.func.isRequired,
  selectDocuments: PropTypes.func.isRequired,
  unselectAllDocuments: PropTypes.func.isRequired,
};

export function mapStateToProps(state, props) {
  return {
    markers: state[props.storeKey].markers,
  };
}

function mapDispatchToProps(dispatch, props) {
  return bindActionCreators(
    { getAndSelectDocument, selectDocuments, unselectAllDocuments },
    wrapDispatch(dispatch, props.storeKey)
  );
}

export default connect(mapStateToProps, mapDispatchToProps, null, { withRef: true })(MapView);
