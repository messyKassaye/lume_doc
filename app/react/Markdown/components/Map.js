import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import Immutable from 'immutable';

import { Map, Markers } from 'app/Map';
import { TemplateLabel } from 'app/Layout';
import {
  getAndSelectDocument,
  selectDocuments,
  unselectAllDocuments,
} from 'app/Library/actions/libraryActions';
import { wrapDispatch } from 'app/Multireducer';

import Loader from 'app/components/Elements/Loader';
import markdownDatasets from '../markdownDatasets';

const renderInfo = marker => (
  <div>
    <TemplateLabel template={marker.properties.entity.template} />
    &nbsp; {marker.properties.entity.title}
  </div>
);

export const MapComponent = props => {
  const { data, classname, scrollZoom, showControls } = props;
  const clickOnMarker = marker => props.getAndSelectDocument(marker.properties.entity.sharedId);
  const clickOnCluster = cluster => {
    props.unselectAllDocuments();
    props.selectDocuments(cluster.map(m => m.properties.entity));
  };

  return (
    <div className={`Map ${classname}`}>
      {data ? (
        <Markers entities={data}>
          {markers => (
            <Map
              markers={markers}
              zoom={1}
              clickOnMarker={clickOnMarker}
              clickOnCluster={clickOnCluster}
              renderPopupInfo={renderInfo}
              cluster
              scrollZoom={scrollZoom === 'true'}
              showControls={showControls === 'true'}
            />
          )}
        </Markers>
      ) : (
        <Loader />
      )}
    </div>
  );
};

MapComponent.defaultProps = {
  classname: '',
  data: null,
  scrollZoom: null,
  showControls: null,
};

MapComponent.propTypes = {
  classname: PropTypes.string,
  data: PropTypes.instanceOf(Immutable.List),
  scrollZoom: PropTypes.string,
  showControls: PropTypes.string,
  getAndSelectDocument: PropTypes.func.isRequired,
  selectDocuments: PropTypes.func.isRequired,
  unselectAllDocuments: PropTypes.func.isRequired,
};

export const mapStateToProps = (state, props) => ({
  data: markdownDatasets.getRows(state, props),
});

export const mapDispatchToProps = dispatch =>
  bindActionCreators(
    { getAndSelectDocument, selectDocuments, unselectAllDocuments },
    wrapDispatch(dispatch, 'library')
  );

export default connect(mapStateToProps, mapDispatchToProps, null)(MapComponent);
