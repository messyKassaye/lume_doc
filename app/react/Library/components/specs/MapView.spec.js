import React from 'react';
import { shallow } from 'enzyme';
import { Markers } from 'app/Map';
import Immutable from 'immutable';

import { MapView, renderInfo } from '../MapView';

describe('MapView', () => {
  let component;
  let props;
  let instance;

  const documents = {
    rows: [
      {
        template: '1',
        metadata: { geo_prop: [{ value: { lat: 1, lon: 2 } }], age: [{ value: '23' }] },
        sharedId: '1',
      },
      { template: '1', metadata: { age: [{ value: '23' }] }, sharedId: '2' },
      {
        template: '2',
        metadata: { other_geo_prop: [{ value: { lat: 3, lon: 4 } }], age: [{ value: '23' }] },
        sharedId: '3',
      },
      {
        template: '2',
        metadata: { other_geo_prop: [{ value: { lat: 1, lon: 2 } }], age: [{ value: '23' }] },
        sharedId: '4',
      },
      { template: '3', metadata: { age: [{ value: '23' }] }, sharedId: '4' },
    ],
  };
  const templates = [
    {
      _id: '1',
      properties: [
        { name: 'geo_prop', type: 'geolocation' },
        { name: 'age', type: 'text' },
      ],
    },
    {
      _id: '2',
      properties: [
        { name: 'other_geo_prop', type: 'geolocation' },
        { name: 'age', type: 'text' },
      ],
    },
    { _id: '3', properties: [{ name: 'age', type: 'text' }] },
  ];

  const render = () => {
    props = {
      markers: Immutable.fromJS(documents),
      storeKey: 'library',
      templates: Immutable.fromJS(templates),
      getAndSelectDocument: jasmine.createSpy('getAndSelectDocument'),
      selectDocuments: jasmine.createSpy('selectDocuments'),
      unselectAllDocuments: jasmine.createSpy('unselectAllDocuments'),
    };
    component = shallow(<MapView {...props} />);
    instance = component.instance();
  };

  beforeEach(render);

  describe('render()', () => {
    it('should render a map', () => {
      const map = component
        .find(Markers)
        .props()
        .children([{ value: 'markers' }]);
      expect(map).toMatchSnapshot();
    });

    it('should pass the entities', () => {
      expect(component.find(Markers).props().entities).toMatchSnapshot();
    });
  });

  describe('clickOnMarker()', () => {
    it('should unselect all documents and select the one in the marker', () => {
      const marker = { latitude: 1, longitude: 2, properties: { entity: documents.rows[2] } };
      instance.clickOnMarker(marker);
      expect(props.getAndSelectDocument).toHaveBeenCalledWith(documents.rows[2].sharedId);
    });
  });

  describe('clickOnCluster()', () => {
    it('should unselect all documents and select the documents in the cluster', () => {
      const cluster = [
        { latitude: 1, longitude: 2, properties: { entity: documents.rows[2] } },
        { latitude: 1, longitude: 2, properties: { entity: documents.rows[3] } },
      ];
      instance.clickOnCluster(cluster);
      expect(props.unselectAllDocuments).toHaveBeenCalled();
      expect(props.selectDocuments).toHaveBeenCalledWith([documents.rows[2], documents.rows[3]]);
    });
  });

  describe('renderPopupInfo', () => {
    it('should render the popup for simple geolocations', () => {
      const marker = {
        properties: {
          entity: {
            template: '5e0fdf2ad6ecbf60be14f094',
            title: 'Uruguay',
          },
          info: '',
        },
        label: 'Geolocation',
      };

      const popup = renderInfo(marker);
      expect(popup).toMatchSnapshot();
    });

    it('should render the popup for inherited geolocations', () => {
      const marker = {
        properties: {
          entity: {
            template: '5e0fde90d6ecbf60be14eca2',
            title: 'Secure investigation and documentation',
          },
          info: 'should not show',
          inherited: true,
          label: 'Bolivia',
        },
        label: 'Country',
      };

      const popup = renderInfo(marker);
      expect(popup).toMatchSnapshot();
    });

    it('should render the popup for non inherited geolocations with info', () => {
      const marker = {
        properties: {
          entity: {
            template: '5e0fde90d6ecbf60be14eca2',
            title: 'Entity with info',
          },
          info: 'Info to Show',
          inherited: false,
        },
        label: 'Entity',
      };

      const popup = renderInfo(marker);
      expect(popup).toMatchSnapshot();
    });
  });
});
