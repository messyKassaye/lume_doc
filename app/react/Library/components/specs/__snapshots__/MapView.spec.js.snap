// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapView render() should pass the entities 1`] = `
Immutable.List [
  Immutable.Map {
    "template": "1",
    "metadata": Immutable.Map {
      "geo_prop": Immutable.List [
        Immutable.Map {
          "value": Immutable.Map {
            "lat": 1,
            "lon": 2,
          },
        },
      ],
      "age": Immutable.List [
        Immutable.Map {
          "value": "23",
        },
      ],
    },
    "sharedId": "1",
  },
  Immutable.Map {
    "template": "1",
    "metadata": Immutable.Map {
      "age": Immutable.List [
        Immutable.Map {
          "value": "23",
        },
      ],
    },
    "sharedId": "2",
  },
  Immutable.Map {
    "template": "2",
    "metadata": Immutable.Map {
      "other_geo_prop": Immutable.List [
        Immutable.Map {
          "value": Immutable.Map {
            "lat": 3,
            "lon": 4,
          },
        },
      ],
      "age": Immutable.List [
        Immutable.Map {
          "value": "23",
        },
      ],
    },
    "sharedId": "3",
  },
  Immutable.Map {
    "template": "2",
    "metadata": Immutable.Map {
      "other_geo_prop": Immutable.List [
        Immutable.Map {
          "value": Immutable.Map {
            "lat": 1,
            "lon": 2,
          },
        },
      ],
      "age": Immutable.List [
        Immutable.Map {
          "value": "23",
        },
      ],
    },
    "sharedId": "4",
  },
  Immutable.Map {
    "template": "3",
    "metadata": Immutable.Map {
      "age": Immutable.List [
        Immutable.Map {
          "value": "23",
        },
      ],
    },
    "sharedId": "4",
  },
]
`;

exports[`MapView render() should render a map 1`] = `
<Map
  autoCenter={true}
  clickOnCluster={[Function]}
  clickOnMarker={[Function]}
  cluster={true}
  height={null}
  hoverOnMarker={[Function]}
  interactiveLayerIds={Array []}
  latitude={null}
  longitude={null}
  markers={
    Array [
      Object {
        "value": "markers",
      },
    ]
  }
  onClick={[Function]}
  renderMarker={null}
  renderPopupInfo={[Function]}
  scrollZoom={true}
  showControls={false}
  width={null}
  zoom={1}
/>
`;

exports[`MapView renderPopupInfo should render the popup for inherited geolocations 1`] = `
<div
  className="popup-container"
>
  <div
    className="template-label"
  >
    <Connect(TemplateLabel)
      template="5e0fde90d6ecbf60be14eca2"
    />
  </div>
  <div
    className="entity-data"
  >
    <div>
      <span
        className="popup-name"
      >
        Secure investigation and documentation
      </span>
      <span
        className="popup-metadata-property"
      >
        (
        Country
        )
      </span>
    </div>
    <div
      className="marker-info"
    >
      <Connect(Icon)
        className="tag-icon"
        icon="tag"
      />
      <span
        className="entity-title"
      >
        Bolivia
      </span>
    </div>
  </div>
</div>
`;

exports[`MapView renderPopupInfo should render the popup for non inherited geolocations with info 1`] = `
<div
  className="popup-container"
>
  <div
    className="template-label"
  >
    <Connect(TemplateLabel)
      template="5e0fde90d6ecbf60be14eca2"
    />
  </div>
  <div
    className="entity-data"
  >
    <div>
      <span
        className="popup-name"
      >
        Entity with info
      </span>
      <span
        className="popup-metadata-property"
      >
        (
        Entity
        )
      </span>
    </div>
    <div
      className="marker-info"
    >
      <Connect(Icon)
        className="tag-icon"
        icon="tag"
      />
      Info to Show
    </div>
  </div>
</div>
`;

exports[`MapView renderPopupInfo should render the popup for simple geolocations 1`] = `
<div
  className="popup-container"
>
  <div
    className="template-label"
  >
    <Connect(TemplateLabel)
      template="5e0fdf2ad6ecbf60be14f094"
    />
  </div>
  <div
    className="entity-data"
  >
    <div>
      <span
        className="popup-name"
      >
        Uruguay
      </span>
      <span
        className="popup-metadata-property"
      >
        (
        Geolocation
        )
      </span>
    </div>
    
  </div>
</div>
`;
